### DISCUSSION TITLE
[UCRT64] Error: junk at end of line, first unrecognized character is `.'

### DISCUSSION THREAD

**Discussion by ghost** [msys2/MINGW-packages#19694] [General]
`example.i` and `example.c` taken from https://swig.org/tutorial.html. I followed the instructions for Python, except I used `python-config --cflags` instead of `-I/usr/local/include/python2.7`. GCC will always give me this error:

```
$ gcc `python-config --cflags` -c example.c example_wrap.c
{standard input}: Assembler messages:
{standard input}:21131: Error: junk at end of line, first unrecognized character is `.'
{standard input}:29381: Error: junk at end of line, first unrecognized character is `.'
```

```
$ gcc -v
Using built-in specs.
COLLECT_GCC=C:\msys64\ucrt64\bin\gcc.exe
COLLECT_LTO_WRAPPER=C:/msys64/ucrt64/bin/../lib/gcc/x86_64-w64-mingw32/13.2.0/lto-wrapper.exe
Target: x86_64-w64-mingw32
Configured with: ../gcc-13.2.0/configure --prefix=/ucrt64 --with-local-prefix=/ucrt64/local --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32 --target=x86_64-w64-mingw32 --with-native-system-header-dir=/ucrt64/include --libexecdir=/ucrt64/lib --enable-bootstrap --enable-checking=release --with-arch=nocona --with-tune=generic --enable-languages=c,lto,c++,fortran,ada,objc,obj-c++,jit --enable-shared --enable-static --enable-libatomic --enable-threads=posix --enable-graphite --enable-fully-dynamic-string --enable-libstdcxx-filesystem-ts --enable-libstdcxx-time --disable-libstdcxx-pch --enable-lto --enable-libgomp --disable-libssp --disable-multilib --disable-rpath --disable-win32-registry --disable-nls --disable-werror --disable-symvers --with-libiconv --with-system-zlib --with-gmp=/ucrt64 --with-mpfr=/ucrt64 --with-mpc=/ucrt64 --with-isl=/ucrt64 --with-pkgversion='Rev3, Built by MSYS2 project' --with-bugurl=https://github.com/msys2/MINGW-packages/issues --with-gnu-as --with-gnu-ld --disable-libstdcxx-debug --with-boot-ldflags=-static-libstdc++ --with-stage1-ldflags=-static-libstdc++
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 13.2.0 (Rev3, Built by MSYS2 project)
```

```
$ python-config --cflags
-IC:/msys64/ucrt64/include/python3.11 -IC:/msys64/ucrt64/include/python3.11  -march=nocona -msahf -mtune=generic -O2 -pipe -Wp,-D_FORTIFY_SOURCE=2 -fstack-protector-strong -O3 -DNDEBUG -g -fwrapv -O3 -Wall
```

**Comment by Biswa96** [comment]
`gcc -c example.c example_wrap.c` worked in my case.

**Reply by ghost** [reply]
How it worked for you is still a mystery for me.

```
$ gcc -c example.c example_wrap.c
example_wrap.c:168:11: fatal error: Python.h: No such file or directory
  168 | # include <Python.h>
      |           ^~~~~~~~~~
compilation terminated.
```

**Reply by Biswa96** [reply]
There is no python.h in example_wrap.c file. Here are the steps I did.
1. Copy-paste the example.c and example.i.
2. `swig -tcl example.i`
3. `gcc -c example.c example_wrap.c`

**Reply by ghost** [reply]
The correct command is `swig -python example.i`.

**Reply by Biswa96** [reply]
With that, your given command works without any error.

```
gcc `python-config --cflags` -c example.c example_wrap.c
```

**Reply by ghost** [reply]
It works with MSYS GCC, not UCRT64 GCC.

**Reply by Biswa96** [reply]
I have used ucrt64 gcc in ucrt64 environment.

**Reply by ghost** [reply]
It doesn't work for me. The problem is from `example_wrap.c` that is generated by `swig`:

```
$ gcc `python-config --cflags` -c example_wrap.c
{standard input}: Assembler messages:
{standard input}:21131: Error: junk at end of line, first unrecognized character is `.'
{standard input}:29381: Error: junk at end of line, first unrecognized character is `.'
```

**Reply by ghost** [reply]
Which version of `swig` are you using? I'm using `mingw-w64-ucrt-x86_64-swig`.

**Reply by Biswa96** [reply]
> Which version of `swig` are you using? I'm using `mingw-w64-ucrt-x86_64-swig`.

Me too.

```
$ swig -version

SWIG Version 4.1.1

Compiled with g++ [x86_64-w64-mingw32]

Configured options: +pcre

Please see https://www.swig.org for reporting bugs and further information
```

**Comment by Biswa96** [comment]
Do these output matche in your case?

```
$ type gcc python-config swig
gcc is hashed (/ucrt64/bin/gcc)
python-config is /ucrt64/bin/python-config
swig is /ucrt64/bin/swig
```

**Reply by ghost** [reply]
The output is same as you.

**Comment by Biswa96** [comment]
Does `gcc $(python-config --cflags) -c example_wrap.c` work in your case?

**Reply by ghost** [reply]
I gives the same error.

**Reply by ghost** [reply]
If I used `-S` instead of `-c` then it successfully generated `example_wrap.s`.

**Reply by ghost** [reply]
I also tried to add `--verbose` to your command and now I'm pretty sure the error message is from the assembler.

**Reply by ghost** [reply]
Strangely enough, if I compile to assembly first then invoke `gcc -c example_wrap.s` then it will successfully compiled the object file.

**Reply by ghost** [reply]
I guess somehow `gcc` passed invalid arguments to `as` when compiling with `gcc -c example_wrap.c` so `as` complained. When compiling to assembly first, this bug is not triggered so it worked as expected. Too bad, even with `--verbose`, I still can't see what `gcc` passed to `as` when compiling with `gcc -c example_wrap.c` to verify my hypothesis.

**Reply by ghost** [reply]
@MehdiChinoune The bug is specific to the UCRT64 GCC. Third party GCC distributions don't have such problem.

Tested: https://github.com/brechtsanders/winlibs_mingw/releases/download/13.2.0posix-17.0.5-11.0.1-ucrt-r3/winlibs-x86_64-posix-seh-gcc-13.2.0-llvm-17.0.5-mingw-w64ucrt-11.0.1-r3.7z

**Reply by ghost** [reply]
One additional information: I'm running Windows 8.1 64 bit.
